# MANE 6399 - Data Science in Manufacturing
## Homework 5 Assignment

### Your Name

In [1]:
# Cell 2 - load data into dataframe df
import pandas as pd

df=pd.read_excel('data-table-B13.XLS',sheet_name='Sheet1')
print(df.columns)
print(df.describe())

Index(['y', 'x1', 'x2', 'x3', 'x4', 'x5', 'x6'], dtype='object')
                 y           x1            x2           x3          x4  \
count    40.000000    40.000000     40.000000     40.00000   40.000000   
mean   3904.000000  1809.925000  19494.600000  29605.77500  174.500000   
std     504.646004   251.950482    940.944419    466.21776   23.436767   
min    3045.000000  1388.000000  17780.000000  28675.00000  133.000000   
25%    3517.500000  1607.750000  18880.000000  29302.50000  155.250000   
50%    3977.000000  1849.500000  19765.000000  29745.00000  179.000000   
75%    4332.500000  2023.750000  20286.250000  29960.00000  193.500000   
max    4833.000000  2239.000000  20740.000000  30250.00000  216.000000   

                x5          x6  
count    40.000000   40.000000  
mean   1651.900000   97.425000  
std      68.895983    4.717059  
min    1522.000000   85.000000  
25%    1592.500000   97.000000  
50%    1668.000000   99.000000  
75%    1709.500000  100.000000  
max 

In [4]:
# Cell 3
# Python code applying regressionDiagnostics function to the Jet Engine Thrust data
#this code extends the Mallows CP example
import numpy as np
import pandas as pd
from itertools import chain, combinations
import statsmodels.formula.api as sm
import statsmodels.tools.tools as smt
import matplotlib.pyplot as plt
# xrange is a Python 2 construct not defined in Python 3
def xrange(x):
    return iter(range(x))
def regressionDiagnostics(X, y):
    n_features = X.shape[1]
    subsets = chain.from_iterable(combinations(xrange(n_features), k+1) for k in xrange(n_features))
    #find full model and perform regression
    for subset in subsets:
        foo=np.array(subset)
        if foo.size == n_features:
            tmpX=smt.add_constant(X.iloc[:,foo])
            lin_reg_full=sm.OLS(y,tmpX).fit()
            N=X.shape[0]
            mse_full=lin_reg_full.mse_resid
    # create arrays to hold results
    modelArray=[]
    cpArray=[]
    pArray=[]
    rSquaredAdj=[]
    mse_res=[]
    aicArray=[]
    bicArray=[]
    #process all models and store results
    subsets = chain.from_iterable(combinations(xrange(n_features), k+1) for k in xrange(n_features))
    for subset in subsets:
        subsetArray=np.array(subset)
        p=subsetArray.size
        tmpX=smt.add_constant(X.iloc[:,subsetArray])
        lin_reg = sm.OLS(y,tmpX).fit()
        p=subsetArray.size
        Cp=(lin_reg.ess/mse_full)-N+2.0*p
        modelArray.append(subset)
        cpArray.append(Cp)
        pArray.append(p)
        rSquaredAdj.append(lin_reg.rsquared_adj)
        mse_res.append(lin_reg.mse_resid)
        aicArray.append(lin_reg.aic)
        bicArray.append(lin_reg.bic)
    cp_df=pd.DataFrame.from_items([('model',modelArray),('p',pArray),('Mallows Cp',cpArray),('R-Squared_Adj',rSquaredAdj),('MSE-residual',mse_res),('AIC',aicArray),('BIC',bicArray)])
    return cp_df
X=df.drop('y',axis=1)
y=df["y"]
#print(X.head(5))
allDiagnostics=regressionDiagnostics(X,y)
print(allDiagnostics.head(10))
print(allDiagnostics.tail(10))
#
print(X.columns)
#save to excel file
writer=pd.ExcelWriter('AllDiagnostics2.xlsx','xlsxwriter')
allDiagnostics.to_excel(writer, sheet_name='Sheet1')
writer.save()

C:\Users\fwn285\AppData\Local\Continuum\anaconda3\envs\env_iise\lib\site-packages\numpy\core\fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


    model  p    Mallows Cp  R-Squared_Adj   MSE-residual         AIC  \
0    (0,)  1  13955.072038       0.989783    2602.012505  430.024969   
1    (1,)  1  13426.822329       0.951424   12370.688038  492.386754   
2    (2,)  1  12154.581541       0.859041   35897.643657  535.000428   
3    (3,)  1  13957.949673       0.989992    2548.797733  429.198433   
4    (4,)  1  10701.520676       0.753528   62768.420628  557.351646   
5    (5,)  1    269.250674      -0.004005  255687.524989  613.531806   
6  (0, 1)  2  13965.599402       0.990143    2510.382639  429.524240   
7  (0, 2)  2  13998.601196       0.992604    1883.601607  418.034260   
8  (0, 3)  2  14000.878651       0.992774    1840.347427  417.105007   
9  (0, 4)  2  13979.575934       0.991185    2244.935699  425.053907   

          BIC  
0  433.402728  
1  495.764513  
2  538.378187  
3  432.576192  
4  560.729405  
5  616.909565  
6  434.590878  
7  423.100898  
8  422.171645  
9  430.120546  
                 model  p    Ma

C:\Users\fwn285\AppData\Local\Continuum\anaconda3\envs\env_iise\lib\site-packages\ipykernel_launcher.py:48: FutureWarning: from_items is deprecated. Please use DataFrame.from_dict(dict(items), ...) instead. DataFrame.from_dict(OrderedDict(items)) may be used to preserve the key order.


# Cell 4
>  Interpret the output of the regressionDiagnostics function. Your answer should include an optimal model for each criterion. From the pool of best models select an overall best model and explain why you selected that model

The interpretation of the regressionDiagnostics data is best performed in Excel. There were five different criterion evaluated: Mallow's Cp, adjusted R^2, MSE-residual, AIC and BIC. Overall Mallow's Cp did not perform well. It indicated the best model was (5) which actually had a negative adjusted R^2 value.

Three of the criterion (adjusted R^2, MSE-residual, and AIC) picked the same model (0,2,3,4,5). The remaining criterion, BIC, selected the model (0,2,4,5). The difference in the two models is neglible and either model is satisfactory.

In [12]:
# Cell 5
#
# Cell 9 - Neural network code
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.neural_network import MLPRegressor
from sklearn.metrics import mean_absolute_error
import matplotlib.pyplot as plt
import pandas as pd
#build the data arrays and scale
qcX=df.drop(df.columns[[0]],axis=1)
#print(qcX)
qcY=df.drop(df.columns[[1,2,3,4,5,6]],axis=1)
#print(qcY)
qc_scaler=MinMaxScaler()
X_scaled=qc_scaler.fit_transform(qcX,qcY)
qcX_train,qcX_test,qcY_train,qcY_test=train_test_split(X_scaled,qcY,test_size=0.25)
#create dataframes for training and testing datasets
foo={'x1':qcX_train[:,0],'x2':qcX_train[:,1],'x3':qcX_train[:,2],'x4':qcX_train[:,3],'x5':qcX_train[:,4],'x6':qcX_train[:,5]} 
nn1_train_df=pd.DataFrame(data=foo)
nn1_train_df['y']=qcY_train.values[:,0]
#print(nn1_train_df)
foo={'x1':qcX_test[:,0],'x2':qcX_test[:,1],'x3':qcX_test[:,2],'x4':qcX_test[:,3],'x5':qcX_test[:,4],'x6':qcX_test[:,5]} 
nn1_test_df=pd.DataFrame(data=foo)
nn1_test_df['y']=qcY_test.values[:,0]
#print(nn1_test_df)
# fit the neural network model
nn1=MLPRegressor(hidden_layer_sizes=(15,),
                activation='logistic',
                solver='sgd',
                max_iter=5000,
                early_stopping=True)
nn1.fit(qcX_train,qcY_train)

#
nn1_train_pred=nn1.predict(qcX_train)
nn1_train_df['Pred_y']=nn1_train_pred[:]
nn1_train_df['resid_y']=nn1_train_df['y']-nn1_train_df['Pred_y']
#
nn1_test_pred=nn1.predict(qcX_test)
nn1_test_df['Pred_y']=nn1_test_pred[:]
nn1_test_df['resid_y']=nn1_test_df['y']-nn1_test_df['Pred_y']
#
# report R2 values
print("The neural network trained for %d epochs"%(nn1.n_iter_))
print("R^2 value for the training set is %f"%(nn1.score(qcX_train,qcY_train)))
print("R^2 value for the test set is %f"%(nn1.score(qcX_test,qcY_test)))
print("MSE for training: %.2f"% np.mean((nn1_train_df['resid_y']) ** 2))
print("MSE for test: %.2f"% np.mean((nn1_test_df['resid_y']) ** 2))
print("Mean absolute error for training: %.2f"%mean_absolute_error(nn1_train_df['y'],nn1_train_df['Pred_y']))
print("Mean absolute error for test: %.2f"%mean_absolute_error(nn1_test_df['y'],nn1_test_df['Pred_y']))

#fig=plt.figure(figsize=(20,8))
#ig.subplots_adjust(hspace=0.8, wspace=0.8)
#ax=fig.add_subplot(131,projection='3d')
#ax.scatter(nn1_train_df['ARL_0'],nn1_train_df['Shift'],nn1_train_df['resid_Lambda'])
#ax.set_xlabel('ARL_0')
#ax.set_ylabel('Shift')
#ax.set_zlabel('residual-Lambda')
#
#ax=fig.add_subplot(132,projection='3d')
#ax.scatter(nn1_train_df['ARL_0'],nn1_train_df['Shift'],nn1_train_df['resid_H'])
#ax.set_xlabel('ARL_0')
#ax.set_ylabel('Shift')
#ax.set_zlabel('residual-H')
#
#ax=fig.add_subplot(133,projection='3d')
#ax.scatter(nn1_train_df['ARL_0'],nn1_train_df['Shift'],nn1_train_df['resid_ARL1'])
#ax.set_xlabel('ARL_0')
#ax.set_ylabel('Shift')
#ax.set_zlabel('residual-ARL1')
#plt.show()


The neural network trained for 32 epochs
R^2 value for the training set is -0.877047
R^2 value for the test set is -0.457386
MSE for training: 425144.31
MSE for test: 451988.95
Mean absolute error for training: 510.79
Mean absolute error for test: 484.09


C:\Users\fwn285\AppData\Local\Continuum\anaconda3\envs\env_iise\lib\site-packages\sklearn\preprocessing\data.py:334: DataConversionWarning: Data with input dtype int64 were all converted to float64 by MinMaxScaler.
  return self.partial_fit(X, y)
C:\Users\fwn285\AppData\Local\Continuum\anaconda3\envs\env_iise\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:1316: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


# Cell 6
> State whether you would recommend the neural network model or the multiple regresson model. What is the basis for your decision? In other words, explain why you made your decision

This neural network performs very poorly on the same dataset that the regression diagnostics was applied to. Notice the negative R^2 scores for the neural network (this requires more investigation; unfortunately, I don't have the time). The MSE-residual was 425144.31 for the training set compared to the best MSE-residual from the regression diagonstics of 694.930.

The reason for the poor performance of this neural network is the configuration of the neural network. To get a much better neural network model, hypertuning of parameters must be implemented. Unfortunately, I did not have the time to try and build the best neural network model possible.